### DISCUSSION TITLE
Animated Webp composition with png problem

### DISCUSSION THREAD

**Discussion by mementobuilds** [ImageMagick/ImageMagick#3771] [Help]
Hi everyone, 

I am trying to overlay a static png (overlay.png) over an animated webp (gradient_background.webp) to get a resulting animated .webp (result.webp). 

There are 2 issues I'm facing: 
1. Where the png is not transparent, it should fully cover the animated webp, but when looking closely at result.webp, there are moving pixels and artefacts noticable in the shape of the canvas of the background webp.
2. The resolution of the result is noticeably lower compared to the input png. 

My command for the composition: `magick convert gradient_background.webp -coalesce null: overlay.png -layers composite result.webp`

I tried several flags like -quality 100 and -define webp:lossless=true or -define webp:method=6 but couldn't fix the issues above. 

Attached a zip with the 3 files mentioned above, as well as a snapshot highlighting the area where pixel movement and artefacts can be noticed in the webp (this part where the png overlays the webp should be static).

[examples.zip](https://github.com/ImageMagick/ImageMagick/files/6577433/examples.zip)

![example](https://user-images.githubusercontent.com/54970488/120350448-311d2080-c329-11eb-8ef1-7f5ff1815b2c.PNG)

Thanks!

**Comment by fmw42** [comment]
In ImageMagick 7, use magick, not magick convert. Please provide your input files or examples that we can use to test. Try adding -layers optimize, though it may not help.

**Reply by fmw42** [reply]
Have you tried the options at https://imagemagick.org/script/webp.php Are you using the latest webp library? What version of ImageMagick and what platform?

**Reply by mementobuilds** [reply]
Thanks @fmw42, examples.zip in the main post above.
Version: ImageMagick 7.0.11-14 Q16 x64 on Windows 10, with webp 1.0.3 delegate lib.

I tried a couple of those options but none changed the output.

**Comment by fmw42** [comment]
I see nothing wrong with your item 1). The moving rainbow pattern is in your gradient.webp animation. With respect to 2), I do not know why the image is slightly less sharp than in the PNG. But I do not know the limitations of webp. I tried adding defines, but it did not make any difference.

`magick gradient_background.webp -coalesce null: overlay.png -layers composite -define lossless=true -define filter-strength=0 result.webp
`

Using IM 7.0.11.14 Q16 Mac OSX Sierra with webp 1.0.3 and viewing with Chrome on my Mac.

**Reply by mementobuilds** [reply]
The issue 1) is that there is pixel movement on the png, kind of like it's shining through it. I attached a short clip (example.mp4) to show what I mean. On this exact example it is not that noticeable but on others that I cannot share here it almost looks like a glitch. 

https://user-images.githubusercontent.com/54970488/120423042-d1a82a80-c393-11eb-8510-a546e7b0fad6.mp4

Regarding issue 2), it surprised me that when exporting to GIF the resolution is much crisper, so it seemed to me that there might be a default lossy compression that's happening with webp's.

**Reply by mementobuilds** [reply]
@fmw42 I'm adding another example and clip with a white overlay png where the glitch is more noticable: 

![image](https://user-images.githubusercontent.com/54970488/120424935-87c14380-c397-11eb-9c60-fd4bbaa0bff0.png)


https://user-images.githubusercontent.com/54970488/120424949-8ee85180-c397-11eb-8ef7-f271bcacc010.mp4

**Reply by fmw42** [reply]
Your first issue looks like some kind of compression or sharpness issue. There are lots of controls, but I really do not know webp enough to say what or how they might affect the result except from the short description. I also do not know enough about how webp is processed in lossless mode to say why it looks to me just a little more blurry than your PNG. Basically, I know too little about webp to comment. Perhaps one of the other users might know more and be more helpful.

**Reply by fmw42** [reply]
I do not see anything there except for the rainbow in the middle.

**Reply by fmw42** [reply]
I will add that if webp is truly lossless, then I would not expect to see shimmering or blurring compared to the PNG at least not where the PNG and webp are supposed to be the same. But is what I see due to your conversion to mp4 or to the magnification you include? I do not see shimmering in the original, but I do see some slight blurring.

**Reply by mementobuilds** [reply]
What you see is how it looks in the animated result.webp, nothing to do with mp4 conversion. It is just more noticable on the white png so I included it. So when you ran that command, the shimmering does not exist? In that case it might be because we have different OS's? I included the white overlay png for testing
![overlay_white](https://user-images.githubusercontent.com/54970488/120426566-86454a80-c39a-11eb-988b-20821ce174b4.png)

But if it's not the case, I might have found an alternative: If I convert to APNG instead of WebP, the quality is much better and no glitches/shimmering. So now I'm trying to convert that APNG to a WebP losslessly. Haven't found a properly working command yet.

**Reply by mementobuilds** [reply]
@fmw42 I additionally found out that there is a bug with the `-quality` flag with webp's: The value 100 outputs a broken (0 bytes) file, whereas the value 99 works. So when using -quality 99, I do get a much better result, without the shimmering and resolution compromises. The file size is much larger though, so will need to find a way to reduce that while keeping a reasonable quality. just reducing the quality value from 99 to something lower quickly starts to add that shimmering effect again.
